### ISSUE TITLE
Enable Code Folding by Default

### ISSUE THREAD

**Issue by rgbkrk** [jupyter/notebook#2235]
Some users that I support love having the [Code folding](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/wiki/Codefolding) extension installed.

![image](https://cloud.githubusercontent.com/assets/836375/23386010/bbdc55ba-fd08-11e6-9669-329658eb36dc.png)

This seems like something we could support out of the box with jupyter notebook instead of having an extension. I'd certainly love having this even for beginners to help deal with unwieldy large notebooks.

**Comment by takluyver** [comment]
I take it the bit about needing CodeMirror master is no longer true?

I'd support enabling code folding by default, but probably leave it to users to configure a shortcut for it.

**Comment by juhasch** [comment]
It is better to link to the generated documentation than to the old Wiki:
http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/codefolding/readme.html

In addition to enabling codefolding in CodeMirror itself, the extension provides saving the folded state in notebook metadata, and an nbconvert preprocessor.

**Comment by takluyver** [comment]
Thanks @juhasch :-)

For putting it in core, I think I'd hold off on storing the state in metadata until we've worked out the 'companion file' question, giving us a place to store metadata that you don't look at in version control.

**Comment by rgbkrk** [comment]
We're going to start using more metadata fields in nteract, albeit on a namespace (nteract). For small fields like these I'm not worried. It's the base64 encoded images that I want to optionally live separately as a relative asset or by absolute URL.

**Comment by takluyver** [comment]
Small metadata is less annoying, but it can be quite changeable, which is annoying for version control.

**Comment by rgbkrk** [comment]
Yeah what I want us to enforce going forward is establishing a schema and version for metadata we introduce: https://github.com/nteract/nteract/issues/1465

**Comment by takluyver** [comment]
I just meant the data itself is changeable, rather than the format. I don't want things like code folding state in version control, any more than I want the last position of my cursor in version control - it makes the diff noisier for little gain.

**Comment by jdriordan** [comment]
@takluyver alternatively for version control with metadata you can use https://github.com/kynan/nbstripout 
I find it works very well

**Comment by takluyver** [comment]
Yep, we've made a variety of tools to make notebooks better with version control. I've done [recombinecm](https://github.com/takluyver/recombinecm) and [nbexplode](https://github.com/takluyver/nbexplode), and [nbdime](https://github.com/jupyter/nbdime) is now recommended. But we try to restrict the changeable metadata in the notebook so that you don't *have* to use any of these tools.

**Comment by jdriordan** [comment]
I understand that people have a variety of different uses, but I feel like
for putting notebook files straight into version control the ship sailed
when outputs, especially images, were saved.

**Comment by takluyver** [comment]
People do put up with it, or commit notebooks without outputs, as you mentioned. I'm keen not to make the situation worse by adding more highly variable metadata, but we have some vague plans taking shape to save two copies of the file - a more version-control friendly file with just inputs and some essential metadata, and a complete file with outputs and more variable metadata.

**Comment by dustinmichels** [comment]
+1 for code folding!

**Comment by rgbkrk** [comment]
To completely skirt the metadata issue, we could consider the code folding without the persistence. It would only be transient.

**Comment by takluyver** [comment]
I'd be fine with enabling this by default without persisting the state for now.

**Comment by BrazilForever11** [comment]
+1 for code folding!

**Comment by sonack** [comment]
+1 code folding is useful

**Comment by agbaezehenry** [comment]
works ok

**Comment by ArmanMirzaei** [comment]
Im on jupyterlab, if the number of code lines within a cell gets more than a few thousands for any reason when working out serious stuff, one would like all the of the code collapsed by default in order to prevent the webbrowser from losing its stability.
any ideas anyone please?
